module/zfs/gzip.c

```diff
@@ -53,16 +54,34 @@ typedef uLongf zlen_t;
 size_t
 gzip_compress(void *s_start, void *d_start, size_t s_len, size_t d_len, int n)
 {
+	int ret;
 	zlen_t dstlen = d_len;
 
 	ASSERT(d_len <= s_len);
 
 	/* check if hardware accelerator can be used */
 	if (qat_dc_use_accel(s_len)) {
-		if (qat_compress(QAT_COMPRESS, s_start,
-		    s_len, d_start, d_len, &dstlen) == CPA_STATUS_SUCCESS)
-			return ((size_t)dstlen);
-		/* if hardware compress fail, do it again with software */
+		/*
+		 * QAT requires some additional scratch space to work with
+		 * to prevent warnings. See qat_compress() for details.
+		 */
+		size_t add_buf_len = d_len;
+		void *add_buf = zio_data_buf_alloc(add_buf_len);
+
+		ret = qat_compress(QAT_COMPRESS, s_start, s_len, d_start,
+		    d_len, add_buf, add_buf_len, &dstlen);
+		zio_data_buf_free(add_buf, add_buf_len);
+
+		if (ret == CPA_STATUS_SUCCESS) {
+			if (dstlen < d_len) {
+				return ((size_t)dstlen);
+			} else {
+				if (d_len == s_len)
```

Comment:
I think this check can be removed.
If the QAT returns success, but the output length (dstLen) is larger or equal to the d_len (87.5% of source length), we just copy the source buffer to dest buffer and return the source length, so ZFS knows the data is not compressed.

Reply:
I did this to match the check below ikn the software implementation. When I didn't have it I ended up getting GPFs.

Reply:
OK, it's different option - we can do the overflow check inside qat_compress function, that is what you've done, that's fine.
Another option I suggested is, not check inside the function, but check the output length in the calling side to decide if we skip the compression (e.g., copy source buffer to dest).
All can be working. Please keep your current implementation if you like, thanks!

Reply:
And, can you please add one comment in the function, saying if it overflow the dest length will be set to the source length ?

Reply:
Sure. Ill add the comment.